### PR TITLE
WIP - Add Filter to return s3 resources that Custodian has limited/no access to

### PR DESCRIPTION
**Pull request by thisisshi**
Returns S3 resources that have `c7n:DeniedActions` that match the `actions` parameter in the `policy.yml`.

`c7n:DeniedActions` is defined by the actions that result in `AccessDenied` when augmenting the S3 resource:

```python
S3_AUGMENT_TABLE = (
    ('get_bucket_location', 'Location', None, None),
    ('get_bucket_tagging', 'Tags', [], 'TagSet'),
    ('get_bucket_policy', 'Policy', None, 'Policy'),
    ('get_bucket_acl', 'Acl', None, None),
    ('get_bucket_replication', 'Replication', None, None),
    ('get_bucket_versioning', 'Versioning', None, None),
    ('get_bucket_website', 'Website', None, None),
    ('get_bucket_logging', 'Logging', None, 'LoggingEnabled'),
    ('get_bucket_notification_configuration', 'Notification', None, None),
    ('get_bucket_lifecycle', 'Lifecycle', None, None),
    #        ('get_bucket_cors', 'Cors'),
)
```

### CODE REVIEW

File: c7n/resources/s3.py [hunk 1]
```diff
@@ -161,6 +161,11 @@ def assemble_bucket(item):
                 methods.append((m, k, default, select))
                 continue
             else:
+                if e.response['Error']['Code'] == 'AccessDenied':
+                    if 'DeniedActions' in b:
+                        b['DeniedActions'].append(m)
```
Comment (kapilt) [inline c7n/resources/s3.py]: re annotation key, we try to keep custodian added fields under the `c7n:` prefix.. ie `c7n:DeniedActions`

File: c7n/resources/s3.py [hunk 2]
```diff
@@ -515,6 +520,41 @@ def __call__(self, b):
         return True
 
 
+@filters.register('no-access')
+class AccessDeniedFilter(Filter):
+    """Find buckets that Custodian can not access.
```
Comment (kapilt) [inline c7n/resources/s3.py]: this one can be removed, its just a value filter on the attribute. ie.

```
filters:
  - type: value
     key: "c7n:DeniedActions"
     op: in
     value_type: swap
     value: get_bucket_location
```

we can add a value type swap_set if you want the multi action matching per the doc string

**Comment by thisisshi** [comment]
@kapilt running a policy like so:

```yaml
policies:
        - name: s3-no-access-test
          resource: s3
          filters:
                  - type: value
                    key: "c7n:DeniedAccess"
                    value_type: swap
                    value: get_bucket_location
                    op: in
```

results in the following 

```
2017-08-07 16:04:08,359: custodian.commands:ERROR Error while executing policy s3-no-access-test, continuing
Traceback (most recent call last):
  File "/cloud-custodian/c7n/commands.py", line 213, in run
    policy()
  File "/cloud-custodian/c7n/policy.py", line 691, in __call__
    resources = PullMode(self).run()
  File "/cloud-custodian/c7n/policy.py", line 305, in run
    resources = self.policy.resource_manager.resources()
  File "/cloud-custodian/c7n/query.py", line 299, in resources
    return self.filter_resources(resources)
  File "/cloud-custodian/c7n/manager.py", line 88, in filter_resources
    resources = f.process(resources, event)
  File "/cloud-custodian/c7n/filters/core.py", line 352, in process
    return super(ValueFilter, self).process(resources, event)
  File "/cloud-custodian/c7n/filters/core.py", line 165, in process
    return list(filter(self, resources))
  File "/cloud-custodian/c7n/filters/core.py", line 339, in __call__
    matched = self.match(i)
  File "/cloud-custodian/c7n/filters/core.py", line 388, in match
    r = self.get_resource_value(self.k, i)
  File "/cloud-custodian/c7n/filters/core.py", line 365, in get_resource_value
    self.expr[k] = jmespath.compile(k)
  File "/cloud-custodian/lib/python2.7/site-packages/jmespath/__init__.py", line 8, in compile
    return parser.Parser().parse(expression)
  File "/cloud-custodian/lib/python2.7/site-packages/jmespath/parser.py", line 88, in parse
    parsed_result = self._do_parse(expression)
  File "/cloud-custodian/lib/python2.7/site-packages/jmespath/parser.py", line 96, in _do_parse
    return self._parse(expression)
  File "/cloud-custodian/lib/python2.7/site-packages/jmespath/parser.py", line 115, in _parse
    "Unexpected token: %s" % t['value'])
ParseError: Unexpected token: :: Parse error at column 3, token ":" (COLON), for expression:
"c7n:DeniedAccess"
    ^
```

It appears that `:` is an illegal character for JMESPath even though the JSON spec itself does not consider `:` illegal. Filtering with the following policy does return the correct resources, however:

```yaml
policies:
        - name: s3-no-access-test
          resource: s3
          filters:
                  - type: no-access
                    actions:
                          - get_bucket_location
```

**Comment by kapilt** [comment]
yeah. there's an ability to make the key a literal string in jmespath which make it valid, but it needs effectively double quoting in the config file. i'm open to alternatives on the annotation prefix, but for some of the state diffing we need to be able to unambigiously determine any custodian annotations, we also need it to behave well with some of the indexing tools we use (elasticsearch has issues with `.`) and .. the alternative i've considered is `c7n-`

**Comment by thisisshi** [comment]
`c7n-DeniedActions` ends up resulting in the same issue:

```
Traceback (most recent call last):
  File "/cloud-custodian/c7n/policy.py", line 305, in run
    resources = self.policy.resource_manager.resources()
  File "/cloud-custodian/c7n/query.py", line 299, in resources
    return self.filter_resources(resources)
  File "/cloud-custodian/c7n/manager.py", line 88, in filter_resources
    resources = f.process(resources, event)
  File "/cloud-custodian/c7n/filters/core.py", line 352, in process
    return super(ValueFilter, self).process(resources, event)
  File "/cloud-custodian/c7n/filters/core.py", line 165, in process
    return list(filter(self, resources))
  File "/cloud-custodian/c7n/filters/core.py", line 339, in __call__
    matched = self.match(i)
  File "/cloud-custodian/c7n/filters/core.py", line 388, in match
    r = self.get_resource_value(self.k, i)
  File "/cloud-custodian/c7n/filters/core.py", line 365, in get_resource_value
    self.expr[k] = jmespath.compile(k)
  File "/cloud-custodian/lib/python2.7/site-packages/jmespath/__init__.py", line 8, in compile
    return parser.Parser().parse(expression)
  File "/cloud-custodian/lib/python2.7/site-packages/jmespath/parser.py", line 88, in parse
    parsed_result = self._do_parse(expression)
  File "/cloud-custodian/lib/python2.7/site-packages/jmespath/parser.py", line 96, in _do_parse
    return self._parse(expression)
  File "/cloud-custodian/lib/python2.7/site-packages/jmespath/parser.py", line 109, in _parse
    self._tokens = list(self.tokenizer)
  File "/cloud-custodian/lib/python2.7/site-packages/jmespath/lexer.py", line 80, in tokenize
    message="Unknown token '%s'" % buff)
LexerError: Bad jmespath expression: Unknown token '-':
c7n-DeniedActions
   ^
```

As it stands right now it seems like the way forward is to fix the way we're resolving the key with JMESPath, which should live inside another pull request

**Comment by kapilt** [comment]
thats fair, re key resolution. to be clear jmespath literals is how we deal with this today.

**Comment by kapilt** [comment]
wrt to literal expressions and double quoting. here's an example

```
$ custodian run -c policies.yml -s out -v
2017-08-12 07:37:16,282: custodian.cli:DEBUG using default region:us-east-1 from boto
2017-08-12 07:37:16,780: custodian.commands:DEBUG Loaded file policies.yml. Contains 1 policies
2017-08-12 07:37:16,859: custodian.output:DEBUG Storing output with <DirectoryOutput to dir:out/s3-no-access-test>
2017-08-12 07:37:16,859: custodian.policy:DEBUG Running policy s3-no-access-test resource: s3 region:us-east-1 c7n:0.8.24.5
2017-08-12 07:37:19,830: custodian.resources.s3:DEBUG Filtered from 5 to 0 s3
2017-08-12 07:37:19,831: custodian.policy:INFO policy: s3-no-access-test resource:s3 region:us-east-1 count:0 time:2.97
2017-08-12 07:37:19,831: custodian.output:DEBUG metric:ResourceCount Count:0 policy:s3-no-access-test restype:s3 scope:policy
(custodian) bash-3.2$ cat policies.yml 
policies:
  - name: s3-no-access-test
    resource: s3
    filters:
      - type: value
        key: '"c7n:DeniedAccess"'
        value_type: swap
        value: get_bucket_location
        op: in
```

**Comment by kapilt** [comment]
i'd like to get this functionality in the next release, which i'm cutting shortly, i'm going to merge it and yank the filter, see my previous comment for how to filter on this.

**Pull request merged**